### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by itmecho**
* Simplify error handling
* Write to CSV Writer during processing instead of storing rows and writing at the end

### CODE REVIEW

**Comment by itmecho** [comment]
Hi @reedloden!

Would you be able to do me a favour and do a test run of this branch? I'd like to make sure it's working somewhere other than my local machine before I merge it in

**Comment by reedloden** [comment]
> Would you be able to do me a favour and do a test run of this branch? I'd like to make sure it's working somewhere other than my local machine before I merge it in

Unfortunately, this branch just hangs for me.

```
./awsinventory -r us-west-2 -l debug
INFO[0000] loading hosted zones
INFO[0000] found 5 hosted zones
INFO[0000] loading route53 records for hosted zone [removed]. (/hostedzone/[removed])
INFO[0000] loading route53 records for hosted zone [removed]. (/hostedzone/[removed])
INFO[0000] loading route53 records for hosted zone [removed]. (/hostedzone/[removed])
INFO[0000] loading route53 records for hosted zone [removed]. (/hostedzone/[removed])
INFO[0000] loading route53 records for hosted zone [removed]. (/hostedzone/[removed])
INFO[0000] found 3 records in hosted zone [removed]. (/hostedzone/[removed])
INFO[0001] found 3 records in hosted zone [removed]. (/hostedzone/[removed])
INFO[0001] found 3 records in hosted zone [removed]. (/hostedzone/[removed])
INFO[0001] found 3 records in hosted zone [removed]. (/hostedzone/[removed])
INFO[0001] found 100 records in hosted zone [removed]. (/hostedzone/[removed])
DEBU[0001] including CloudFront service
DEBU[0001] including IAM service
DEBU[0001] including CodeCommit service
DEBU[0001] including DynamoDB service
DEBU[0001] including EBS service
DEBU[0001] including EC2 service
DEBU[0001] including ECR service
DEBU[0001] including ECS service
DEBU[0001] including ElastiCache service
DEBU[0001] including Elasticsearch service
DEBU[0001] including ELB service
DEBU[0001] including ELB v2 service
DEBU[0001] including KMS service
INFO[0001] loading data                                  region=global service=iam
INFO[0001] loading data                                  region=us-west-2 service=dynamodb
INFO[0001] loading data                                  region=us-west-2 service=kms
INFO[0001] loading data                                  region=us-west-2 service=ec2
INFO[0001] loading data                                  region=us-west-2 service=es
INFO[0001] loading data                                  region=us-west-2 service=ebs
INFO[0001] loading data                                  region=us-west-2 service=ecs
INFO[0001] loading data                                  region=us-west-2 service=ecr
DEBU[0001] including Lambda service
DEBU[0001] including RDS service
DEBU[0001] including S3 service
DEBU[0001] including SQS service
INFO[0001] loading data                                  region=us-west-2 service=sqs
INFO[0001] loading data                                  region=us-west-2 service=elb
INFO[0001] loading data                                  region=us-west-2 service=elasticache
INFO[0001] loading data                                  region=us-west-2 service=lambda
INFO[0001] loading data                                  region=global service=cloudfront
INFO[0001] loading data                                  region=us-west-2 service=elbv2
INFO[0001] loading data                                  region=us-west-2 service=s3
INFO[0001] loading data                                  region=us-west-2 service=codecommit
INFO[0001] loading data                                  region=us-west-2 service=rds
INFO[0001] processing data                               region=us-west-2 service=ecr
INFO[0001] finished processing data                      region=us-west-2 service=ecr
INFO[0001] processing data                               region=us-west-2 service=es
INFO[0001] processing data                               region=us-west-2 service=kms
INFO[0001] processing data                               region=us-west-2 service=codecommit
INFO[0001] processing data                               region=us-west-2 service=dynamodb
INFO[0001] processing data                               region=us-west-2 service=sqs
INFO[0001] no data found; bailing early                  region=us-west-2 service=ecs
INFO[0001] processing data                               region=us-west-2 service=s3
INFO[0001] processing data                               region=us-west-2 service=elasticache
INFO[0001] finished processing data                      region=us-west-2 service=codecommit
INFO[0001] processing data                               region=us-west-2 service=elbv2
INFO[0001] finished processing data                      region=us-west-2 service=elbv2
INFO[0001] processing data                               region=us-west-2 service=ebs
INFO[0001] finished processing data                      region=us-west-2 service=ebs
INFO[0001] finished processing data                      region=us-west-2 service=dynamodb
INFO[0001] processing data                               region=us-west-2 service=elb
INFO[0001] processing data                               region=us-west-2 service=ec2
INFO[0001] finished processing data                      region=us-west-2 service=ec2
INFO[0001] processing data                               region=us-west-2 service=rds
INFO[0002] processing data                               region=global service=iam
INFO[0002] processing data                               region=us-west-2 service=lambda
INFO[0002] processing data                               region=global service=cloudfront
```

Don't have time to look deeper into this right now, but let me know if you need me to try anything else.

File: cmd/awsinventory/awsinventory.go [hunk 1]
```diff
@@ -61,10 +59,11 @@ func main() {
 
 	// Write stored rows to csv inventory
 	var count int
-	awsData.MapRows(func(row inventory.Row) error {
-		count++
-		return csv.WriteRow(row)
-	})
+	awsData.Load(regions, services, nil)
+	// awsData.Load(regions, services, func(row inventory.Row) error {
```
Comment (reedloden) [inline cmd/awsinventory/awsinventory.go]: Please remove if it is no longer needed.

Comment (itmecho) [reply cmd/awsinventory/awsinventory.go]: Sorry about that! I've switched that back now and added proper handling for that situation. That should only be set to `nil` during testing when we don't care about the rows, for example, when testing that an error is returned.

Comment (itmecho) [reply cmd/awsinventory/awsinventory.go]: The reason I didn't notice the hanging is that the rows channel can only hold 100 rows before blocking and the account I tested against had less than 100 resources in it :roll_eyes:

**Comment by reedloden** [comment]
@itmecho Tested against multiple accounts with multiple regions. All good with 2857 rows. :-)

**Comment by itmecho** [comment]
Thanks for that! I've got an S3 performance improvement to merge after this as well

**Pull request merged**